design/two-stage-backup-api-proposal.md

```diff
@@ -0,0 +1,149 @@
+|                    |
++----------+---------+
+           |
+           |
+           | < 2 >
+           |
+           v
++----------+---------+
+|                    |
+|    Post-Hook       |
+|                    |
++----------+---------+
+           |
+           | < 3 >
+           |
+           v
+ +---------+---------+
+ |                   |
+ |  UploadSnapshot   |
+ |   (Background)    |
+ +-------------------+
+
+
+```
+
+## Detailed Design
+
+Velero has a pretty neat feature to maintain the application consistent backups/snapshot called `Pre-hooks` and `Post-Hooks`. However, there's a small change in Velero which needs to be addressed in order to achieve minimum downtime for an application or more consistent application backups. Existing backup flow for a typical stateful application looks like: `Pod` -> `Pre-Hook` --> `PVC` -> `PV` -> `Post-Hook`. This is a mechanism which is followed by Velero block store plugins. In most of the plugins, `CreateSnapshot` method of `VolumeSnapshotter` interface implements logic to upload the snapshot and this is a blocking call for `Post-Hook`. This will introduce more time and will keep application frozen for time equals to snapshot upload time.
+
+
+To solve the above problem, we can split existing backup flow and introduce one more method called `UploadSnapshot` to `VolumeSnapshotter` interface, which will execute once `Post-Hook` is executed and will keep running in background. This will reduce the time between `Pre-Hook` and `Post-Hook` significantly as well as it will not impact application consistency or downtime.
+
+### Implementation
+
+This change should be implemented in both Velero as well as Velero plugins. However, Velero plugins can continue to rely on current implementation. Only way for Velero plugins to get benifitted from this approach is to implement `UploadSnapshot` method.
+
+Volumesnapshotter is designed for snapshot, this can be either cloud snapshot or local snapshot. As of now, most plugin use volumesnapshotlocations to decide the snapshot type.
+We can add new config parameter in velero command line to define snapshot type, through this, velero will decide whether to invoke `UploadSnapshot` or not.
+
+**Implementation plan for Velero:**
+
+- Add `UploadSnapshot` method to `VolumeSnapshotter` interface.
+- Generate GRPC code using codegen.
+- Implement `UploadSnapshot` method in client and server.
+- Call `UploadSnapshot` method as soon as `Post-Hook` is executed.
+- If Velero plugin doesn't implement `UploadSnapshot` method then handle `Unimplemented` error.
+
+**Implementation plan for Velero Plugin:**
+
+- Update vendor with latest Velero code and make sure that GRPC generated code is present in vendor.
+- Move the logic of upload backup to separate `UploadSnapshot` method.
+
+**Method signature:**
+
+```go
+
+func UploadSnapshot(volumeID string) error {
+...
+}
+
+```
+
+```Note: Backup progress should not be updated until `UploadSnapshot` is returned.```
```

Comment:
> I think for now we probably don't want to add any new backup phases

Right. We should only use existing phases since we are not adding more to the current backup workflow.